design-proposals/vm-pool.md

```diff
@@ -0,0 +1,194 @@
+
+## Goals
+
+* VM pool abstraction capable of managing replication of stateful VMs at scale
+* Automated rollout of spec changes across and other updates across a pool of stateful VMs
+* Automated and manual scale-out and scale-in VMs associated with a pool
+* Autohealing (delete and replace) of VMs in a pool not passing health checks
+* Ability to specify unique secrets and configMap data per VM within a VMPool
+
+## Non-Goals
+
+* Not designing a pool abstraction capable of managing multiple VM shapes within a single pool. A VMPool consists only of VMs replicated from a single config.
+
+## Terms
+
+* **VirtualMachine [VM]** - Refers specifically to the KubeVirt VirtualMachine api object. 
+* **VirtualMachinePool [VMPool]** - KubeVirt API for scaling out/in replicas of KubeVirt VirtualMachines
+* **VirtualMachineConfig [VMConfig]** - KubeVirt API representing a VirtualMachine spec which is referenced by a VirtualMachinePool.
+* **Scale-out and Scale-in** - Terms to describe the action of adding/removing to the count of VMs within a VMPool.
+* **Detach VM** - The process of manually separating a VirtualMachine from a VirtualMachinePool.
+* **Disruption Strategy** - The policy used to define how a VMPool both selects VMs during scale-in and update.
+
+## User Stories
+
+* As a cluster user, I want to automate batch rollout of changes (CPU/Memory/Disk/PubSSHKeys/etc…) across a fleet of VMs.
+* As a cluster user managing fleets of VMs I want to automate scale out of VM instances based on utilization
+* As a cluster user managing fleets of VMs I want to automate scale-in of identical VM instances to optimize cluster resource consumption
+* As a user transitioning workloads to KubeVirt I want to use similar management patterns provided by existing Iaas platforms (AWS, Azure, GCP)
+* As a cluster admin managing nested Kubernetes clusters on top of KubeVirt VMs, I want the ability to elastically scale the underlying KubeVirt VM infrastructure.
+* As a SRE managing the availability of VM fleets I want to automate VM recovery by auto detecting and deleting misbehaving VMs and having the platform spin up fresh new instances as a replacement.
+* As a pool user/manager I want to remove a VM from the pool without modifying it for debugging. The missing VM can be replaced by the pool.
+
+# Design
+
+The VMPool design introduces two new APIs represented as CRDs, the VirtualMachinePool (VMPool) object and the VirtualMachineConfig (VMConfig) object. For the most part, a VMConfig object is simply a VM object without a status section that is used by a VMPool to stamp out multiple replicas.
```

Comment:
Why did you choose to have a separate VMConfig object instead of embedding the VMSpec in the VMPool resource? It's usually pretty handy to have everything in one place for inspecting and replicating a manifest with similar configs.

Reply:
This is explained in some of the next sections.

Reply:
I like that change as it allows to easily create different pools from a single base config.

Reply:
> Why did you choose to have a separate VMConfig object instead of embedding the VMSpec in the VMPool resource? It's usually pretty handy to have everything in one place for inspecting and replicating a manifest with similar configs.

a couple primary reasons off the top of my head.

- it gives VMConfigs a 1 to many relationship to VMPools. Meaning one VMConfig can be assigned to multiple VMPools. 
- it allows people to plan and stage VMPool updates. For example, someone can create different VMConfig versions. Roll out the new VMCOnfig in some dev/staging environments, then roll it out to Prod. Rolling back is as simple are assigning the previous VMConfig to the VMPool.

Reply:
@acardace i'm actually leaning towards your suggestion now and revised the document to embed the VMSpec into the VMPool.